0005-ordered-quantification.md

```diff
@@ -0,0 +1,661 @@
+
+A sequence comprehension has identical syntax to a set comprehension, except that it begins with
+`seq` rather than `set`. Therefore its syntax is:
+
+```
+SeqComprehension ::=
+  "seq" 
+  QuantifierDomain
+  [ :: Expression ]
+```
+
+Sequence comprehensions are likely to be easier to work with than `foreach` loops, and Dafny users should be
+encouraged to use them instead where possible. Many high-order operations on sequences can be expressed directly
+using sequence comprehensions:
+
+```dafny
+// Filtering to optional values
+var s: seq<Option<T>> := ...;
+var filtered: seq<T> := seq o <- s | o.Some? :: o.value;
+
+// Map and then flatten (a.k.a. "FlatMap")
+var c: seq<S> := ...;
+var f: S -> seq<T> := ...;
+var flatMapped: seq<T> := seq x <- c, y <- f(x) :: y;
+
+// Intersection of sequences
+var a := [4, 1, 3, 5];
+var b := [3, 1, 7];
+assert (seq x <- a | x in b) == [1, 3];
+assert (seq x <- b | x in a) == [3, 1];
+```
+
+Since sequence comprehensions are expressions rather than statements, they carry an additional restriction
+when compiled: their ordering must be fully deterministic. If `s` is a `set<int>`, for example,
+`seq x <- s` would be rejected in compiled code.
+This is very similar to the restriction on `:|` let-such-that expressions, which is not relevant for equivalent
+`:|` assign-such-that statements.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As mentioned in the guide-level explanation, `foreach` loops and sequence comprehensions are both able to
+borrow concepts and implementation substantially from other features. Parsing, resolving, verifying, and compiling
+quantifier domains is already a mature aspect of the Dafny implementation. The most significant implementation burden
+is ensuring that the required quantification ordering is threaded through the stack, 
+and checked to be deterministic when used in compiled sequence comprehensions. 
+
+## Verification
+
+The optimal encoding of sequence comprehensions to Boogie is something of an open question:
+set comprehensions are encoded as their indicator predicates, but there is no such direct representation
+for sequence comprehensions as defined here. The initial encoding will likely be with a function,
+similar to the `Seq#Create` function used for seq constructions, that accepts a source sequence and indicator predicate.
+The useful axioms about these values will fall into three categories, sketched below for the simple
+case of comprehensions with a single variable bound from a sequence:
+
+`var S := seq x <- C | P(x) :: Q(x);`
+
+1. Membership: `forall x :: x in C && P(x) ==> Q(x) in S`
+2. Cardinality: `|S| <= |C|`
+3. Ordering: `forall i, j | 0 <= i < j < |S| :: 
+      exists i', j' | 0 <= i' < j' < |C| :: 
+        Q(C[i']) == S[i] && Q(C[j']) == S[j]`
+
+The translation of `foreach` loops should be reducible to sequence comprehensions
+and other existing features for loops in general,
```

Comment:
Sequence comprehensions are more difficult to formalize for the verifier than `foreach` loops, so I don't suggest formalizing `foreach` in terms of sequence comprehensions. I'm thinking that the verifier would handle `foreach` loops like this:
* Check the loop invariant initially (as usual for a loop).
* Inside the loop, play havoc with the loop assignment targets (as usual for a loop) and the bound variables (as is also done in the `for` loop). Then, assume the loop invariant and assume the condition that you get by replacing every `<-` with an `in` in the `foreach` loop header. Then, check the body and check that the body maintains the loop invariant.
* After the loop, assume the loop invariant (as usual for a loop). Here's what I don't know what to do with: one also needs to assume some condition that's analogous to "the negation of the loop guard". I don't know what that condition is here.